// ng-zorro-antd/esm5/notification/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
export { NZ_NOTIFICATION_DEFAULT_CONFIG, NZ_NOTIFICATION_CONFIG, NZ_NOTIFICATION_DEFAULT_CONFIG_PROVIDER } from './nz-notification-config';
export { NzNotificationComponent } from './nz-notification.component';
export { NzNotificationModule } from './nz-notification.module';
export {} from './nz-notification.definitions';
export { NzNotificationService } from './nz-notification.service';
export { NzNotificationServiceModule } from './nz-notification.service.module';
export { NzNotificationContainerComponent } from './nz-notification-container.component';
