// ng-zorro-antd/esm2015/result/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
export { NzResultModule } from './nz-result.module';
export { NzResultComponent } from './nz-result.component';
export { NzResultTitleDirective, NzResultSubtitleDirective, NzResultIconDirective, NzResultContentDirective, NzResultExtraDirective } from './nz-result-cells';
// Making these partial components not visible to users but comprehensive to ng-packagr.
export { NzResultNotFoundComponent as θNzResultNotFoundComponent } from './partial/not-found';
export { NzResultServerErrorComponent as θNzResultServerErrorComponent } from './partial/server-error.component';
export { NzResultUnauthorizedComponent as θNzResultUnauthorizedComponent } from './partial/unauthorized';
