// ng-zorro-antd/esm5/result/nz-result-cells.js
/**
 * @fileoverview added by tsickle
 * Generated from: nz-result-cells.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive } from '@angular/core';
var NzResultTitleDirective = /** @class */ (function () {
    function NzResultTitleDirective() {
    }
    NzResultTitleDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'div[nz-result-title]',
                    exportAs: 'nzResultTitle',
                    host: {
                        class: 'ant-result-title'
                    }
                },] }
    ];
    return NzResultTitleDirective;
}());
export { NzResultTitleDirective };
var NzResultSubtitleDirective = /** @class */ (function () {
    function NzResultSubtitleDirective() {
    }
    NzResultSubtitleDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'div[nz-result-subtitle]',
                    exportAs: 'nzResultSubtitle',
                    host: {
                        class: 'ant-result-subtitle'
                    }
                },] }
    ];
    return NzResultSubtitleDirective;
}());
export { NzResultSubtitleDirective };
var NzResultIconDirective = /** @class */ (function () {
    function NzResultIconDirective() {
    }
    NzResultIconDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'i[nz-result-icon], div[nz-result-icon]',
                    exportAs: 'nzResultIcon'
                },] }
    ];
    return NzResultIconDirective;
}());
export { NzResultIconDirective };
var NzResultContentDirective = /** @class */ (function () {
    function NzResultContentDirective() {
    }
    NzResultContentDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'div[nz-result-content]',
                    exportAs: 'nzResultContent',
                    host: {
                        class: 'ant-result-content'
                    }
                },] }
    ];
    return NzResultContentDirective;
}());
export { NzResultContentDirective };
var NzResultExtraDirective = /** @class */ (function () {
    function NzResultExtraDirective() {
    }
    NzResultExtraDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'div[nz-result-extra]',
                    exportAs: 'nzResultExtra',
                    host: {
                        class: 'ant-result-extra'
                    }
                },] }
    ];
    return NzResultExtraDirective;
}());
export { NzResultExtraDirective };
